### ISSUE TITLE
[java] False positive for rule Unused Assignment

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2999]
**Affects PMD Version:** 
6.30.0
**Description:**

False positive for variable that is pre-incremented and its value used in the expression.

The rule configuration is as follows:

```xml
<rule ref="category/java/bestpractices.xml/UnusedAssignment">
    <properties>
      <property name="checkUnusedPrefixIncrement"
                value="true" />
      <property name="reportUnusedVariables"
                value="true" />
      <property name="violationSuppressXPath" 
        value="./ancestor::MethodDeclaration[contains(@Name, 'main')]"/>
    </properties>
  </rule>
  
```

**Code Sample demonstrating the issue:**

```java
package javapuzzles;

public enum AddNum {
  ;

  public static void main(String... args) {
    System.out.println(computeSum());
    System.out.println(computeSumAgain());
    System.out.println(computeSummer());
  }

  private static int computeSum() {
    int sum = 0;
    int mod = 1;
    sum += mod * ++mod / 2;
    return sum;
  }

  private static int computeSumAgain() {
    int sum = 0;
    int mod = 1;
    sum += mod * (++mod / 2);
    return sum;
  }

  private static int computeSummer() {
    int sum = 0;
    int mod = 1;
    sum += (mod * ++mod) / 2;
    return sum;
  }
}

```
**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by oowekyala** [comment]
These are not FPs. You can replace `++mod` with `(mod + 1)` in all cases

**Comment by Fernal73** [comment]
What does the rule report?
"Report expressions like ++i that may be replaced with (i + 1)"

I'd expect it to report a violation if the value assigned to mod by the prefixed increment is not used. In this case, it is.

**Comment by oowekyala** [comment]
> I'd expect it to report a violation if the value assigned to mod by the prefixed increment is not used. In this case, it is.

The variable is not used after the increment. This would be like saying `mod * (mod = mod + 1)` is ok because the value of mod is used, as it's the value of the assignment expression.

**Comment by Fernal73** [comment]
I don't see it as the variable value not used post the increment. It is
used in the multiplication that happens. And its result in the division.

**Comment by Fernal73** [comment]
> > I'd expect it to report a violation if the value assigned to mod by the prefixed increment is not used. In this case, it is.
> 
> This would be like saying `mod * (mod = mod + 1)` is ok because the value of mod is used, as it's the value of the assignment expression.

That's a [different rule---that too, in Checkstyle that prohibits inner assignments](https://checkstyle.sourceforge.io/config_coding.html#InnerAssignment).

**Comment by Fernal73** [comment]
You're saying that (++mod) is the same as (mod = mod + 1) for the purposes of this rule as in the latter case the return value of the inner assignment is used, not the assigned value. Am I correct? 

The following is the updated code sample and the byte code for the different ways of expressing the same method:


```java
package javapuzzles;

public enum AddNum {
  ;

  public static void main(String... args) {
    System.out.println(computeSum());
    System.out.println(computeSumAgain());
    System.out.println(computeSummer());
    System.out.println(computeSumExpanded());
    System.out.println(computeSumSub());
  }

  private static int computeSum() {
    int sum = 0;
    int mod = 1;
    sum += mod * ++mod / 2;
    return sum;
  }

  private static int computeSumExpanded() {
    int sum = 0;
    int mod = 1;
    int b = mod + 1;
    sum += mod * b / 2;
    return sum;
  }
  
  private static int computeSumSub() {
    int sum = 0;
    int mod = 1;
    sum += mod * (mod = mod + 1)/ 2;
    return sum;
  }

  private static int computeSumAgain() {
    int sum = 0;
    int mod = 1;
    sum += mod * (++mod / 2);
    return sum;
  }

  private static int computeSummer() {
    int sum = 0;
    int mod = 1;
    sum += (mod * ++mod) / 2;
    return sum;
  }
}

```


```assembly
private static int computeSum();
    Code:
       0: iconst_0
       1: istore_0
       2: iconst_1
       3: istore_1
       4: iload_0
       5: iload_1
       6: iinc          1, 1
       9: iload_1
      10: imul
      11: iconst_2
      12: idiv
      13: iadd
      14: istore_0
      15: iload_0
      16: ireturn

  private static int computeSumExpanded();
    Code:
       0: iconst_0
       1: istore_0
       2: iconst_1
       3: istore_1
       4: iload_1
       5: iconst_1
       6: iadd
       7: istore_2
       8: iload_0
       9: iload_1
      10: iload_2
      11: imul
      12: iconst_2
      13: idiv
      14: iadd
      15: istore_0
      16: iload_0
      17: ireturn

  private static int computeSumSub();
    Code:
       0: iconst_0
       1: istore_0
       2: iconst_1
       3: istore_1
       4: iload_0
       5: iload_1
       6: iload_1
       7: iconst_1
       8: iadd
       9: dup
      10: istore_1
      11: imul
      12: iconst_2
      13: idiv
      14: iadd
      15: istore_0
      16: iload_0
      17: ireturn

  private static int computeSumAgain();
    Code:
       0: iconst_0
       1: istore_0
       2: iconst_1
       3: istore_1
       4: iload_0
       5: iload_1
       6: iinc          1, 1
       9: iload_1
      10: iconst_2
      11: idiv
      12: imul
      13: iadd
      14: istore_0
      15: iload_0
      16: ireturn

  private static int computeSummer();
    Code:
       0: iconst_0
       1: istore_0
       2: iconst_1
       3: istore_1
       4: iload_0
       5: iload_1
       6: iinc          1, 1
       9: iload_1
      10: imul
      11: iconst_2
      12: idiv
      13: iadd
      14: istore_0
      15: iload_0
      16: ireturn
```

**Comment by Fernal73** [comment]
> You're saying that (++mod) is the same as (mod = mod + 1) for the purposes of this rule as in the latter case the return value of the inner assignment is used, not the variable. Am I correct?
> 
> The following is the updated code sample and the byte code for the different ways of expressing the same method:
> 
> ```java
> package javapuzzles;
> 
> public enum AddNum {
>   ;
> 
>   public static void main(String... args) {
>     System.out.println(computeSum());
>     System.out.println(computeSumAgain());
>     System.out.println(computeSummer());
>     System.out.println(computeSumExpanded());
>     System.out.println(computeSumSub());
> System.out.println(computeSumPlus());
> 
>   }
> 
>   private static int computeSum() {
>     int sum = 0;
>     int mod = 1;
>     sum += mod * ++mod / 2;
>     return sum;
>   }
> 
> private static int computeSumPlus() {
>     int sum = 0;
>     int mod = 1;
>     sum += mod * (mod + 1) / 2;
>     return sum;
>   }
> 
>   private static int computeSumExpanded() {
>     int sum = 0;
>     int mod = 1;
>     int b = mod + 1;
>     sum += mod * b / 2;
>     return sum;
>   }
> 
>   
>   private static int computeSumSub() {
>     int sum = 0;
>     int mod = 1;
>     sum += mod * (mod = mod + 1)/ 2;
>     return sum;
>   }
> 
>   private static int computeSumAgain() {
>     int sum = 0;
>     int mod = 1;
>     sum += mod * (++mod / 2);
>     return sum;
>   }
> 
>   private static int computeSummer() {
>     int sum = 0;
>     int mod = 1;
>     sum += (mod * ++mod) / 2;
>     return sum;
>   }
> }
> ```
> 
> ```assembly
> private static int computeSum();
>     Code:
>        0: iconst_0
>        1: istore_0
>        2: iconst_1
>        3: istore_1
>        4: iload_0
>        5: iload_1
>        6: iinc          1, 1
>        9: iload_1
>       10: imul
>       11: iconst_2
>       12: idiv
>       13: iadd
>       14: istore_0
>       15: iload_0
>       16: ireturn
> 
> private static int computeSumPlus();
>     Code:
>        0: iconst_0
>        1: istore_0
>        2: iconst_1
>        3: istore_1
>        4: iload_0
>        5: iload_1
>        6: iload_1
>        7: iconst_1
>        8: iadd
>        9: imul
>       10: iconst_2
>       11: idiv
>       12: iadd
>       13: istore_0
>       14: iload_0
>       15: ireturn
> 
>   private static int computeSumExpanded();
>     Code:
>        0: iconst_0
>        1: istore_0
>        2: iconst_1
>        3: istore_1
>        4: iload_1
>        5: iconst_1
>        6: iadd
>        7: istore_2
>        8: iload_0
>        9: iload_1
>       10: iload_2
>       11: imul
>       12: iconst_2
>       13: idiv
>       14: iadd
>       15: istore_0
>       16: iload_0
>       17: ireturn
> 
>   private static int computeSumSub();
>     Code:
>        0: iconst_0
>        1: istore_0
>        2: iconst_1
>        3: istore_1
>        4: iload_0
>        5: iload_1
>        6: iload_1
>        7: iconst_1
>        8: iadd
>        9: dup
>       10: istore_1
>       11: imul
>       12: iconst_2
>       13: idiv
>       14: iadd
>       15: istore_0
>       16: iload_0
>       17: ireturn
> 
>   private static int computeSumAgain();
>     Code:
>        0: iconst_0
>        1: istore_0
>        2: iconst_1
>        3: istore_1
>        4: iload_0
>        5: iload_1
>        6: iinc          1, 1
>        9: iload_1
>       10: iconst_2
>       11: idiv
>       12: imul
>       13: iadd
>       14: istore_0
>       15: iload_0
>       16: ireturn
> 
>   private static int computeSummer();
>     Code:
>        0: iconst_0
>        1: istore_0
>        2: iconst_1
>        3: istore_1
>        4: iload_0
>        5: iload_1
>        6: iinc          1, 1
>        9: iload_1
>       10: imul
>       11: iconst_2
>       12: idiv
>       13: iadd
>       14: istore_0
>       15: iload_0
>       16: ireturn
> ```
> 
> https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.15.1
> 
> The documentation needs to be updated appropriately.

From what I can see above, the bytecode is most efficient when `mod * (mod + 1)` is used and `mod` is not referred to in any further statements.

Semantically, it still doesn't seem right, though.

**Comment by oowekyala** [comment]
When you write `++mod`, the value `mod + 1` is stored in `mod`. The problem in your code is that the variable mod is never used after that, so that the *store* is useless. This is why it's reported: it can be replaced by `mod + 1` and not do any side effect.

For example in `++mod * mod` there is no problem, because the increment is observed by the second load of `mod`. This has nothing to do with the fact that this is an inner assignment. PMD also has a separate rule about that.

I don't think there's anything to do about the documentation. I think you misunderstand the purpose of this property/rule.

**Comment by Fernal73** [comment]
I understood it as whether the updated value is used or not---which had nothing to do with the byte code.

**Comment by oowekyala** [comment]
The updated value is `mod + 1`, and the side-effect is not used. So it's just obscure to use `++mod` instead of `mod+1`. If you don't like this just don't set the property...

**Comment by Fernal73** [comment]
As far as I can see, the byte code is more efficient using (mod + 1).
The property was unset before in my configuration. I set it to see what it
will produce.
